File: scripts/converters/codex.py
"""Codex CLI converter — bundle N SKILL.md files into a single AGENTS.md.

Codex CLI reads `AGENTS.md` at the project root as a single context file
(there's no per-skill activation — the LLM sees everything at once). Our
job is to consolidate multiple installed skills into that file WITHOUT
clobbering content the user already wrote.

Strategy: managed section with HTML-comment markers.

    <!-- skillforge:start -->
    # SkillForge skills
    ... generated content ...
    <!-- skillforge:end -->

    User content lives outside the markers and survives re-runs.

Merge semantics on re-convert:
    - If AGENTS.md exists and already has a managed section → replace the
      section content, leave everything else alone.
    - If AGENTS.md exists with no markers → append the managed section at
      the end, preceded by a blank line.
    - If AGENTS.md doesn't exist → create a new file with just the
      managed section.

Per-skill format inside the managed section:
    ## <skill id>
    _Activate when: <description>_

    <skill body>

    ---
    (next skill)

Scripts/references dirs are still non-portable — Codex's single-file
model can't carry executable assets. The converter emits a warning for
each skipped asset dir, same as the Cursor converter.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import List, Tuple

from .cursor import _parse_frontmatter  # reuse the frontmatter parser


START_MARKER = "<!-- skillforge:start -->"
END_MARKER = "<!-- skillforge:end -->"


@dataclass
class SkillInput:
    """Input for one skill passed to bundle_skills."""
    id: str
    skill_md_text: str
    has_scripts: bool = False
    has_references: bool = False


@dataclass
class BundleResult:
    """Full AGENTS.md content to write, plus any warnings."""
    content: str = ""
    warnings: List[str] = field(default_factory=list)


def _render_skill_section(skill: SkillInput) -> Tuple[str, List[str]]:
    """Render one skill as a `## <id>` section. Returns (markdown, warnings)."""
    warnings: List[str] = []
    fm, body = _parse_frontmatter(skill.skill_md_text)
    description = fm.get("description", "").strip()
    if not description:
        description = skill.id
        warnings.append(
            f"{skill.id}: no description in SKILL.md frontmatter — "
            f"Codex section uses id as fallback activation hint"
        )
    if skill.has_scripts:
        warnings.append(
            f"{skill.id}: scripts/ not portable to Codex — section "
            f"contains only the markdown body"
        )
    if skill.has_references:
        warnings.append(
            f"{skill.id}: references/ not portable to Codex — section "
            f"contains only the markdown body"
        )
    body_stripped = body.strip("\n")
    section = (
        f"## {skill.id}\n"
        f"_Activate when: {description}_\n\n"
        f"{body_stripped}\n"
    )
    return section, warnings


def _render_managed_section(skills: List[SkillInput]) -> Tuple[str, List[str]]:
    """Build the full managed block bracketed by markers."""
    all_warnings: List[str] = []
    sections: List[str] = []
    for skill in skills:
        section, warnings = _render_skill_section(skill)
        all_warnings.extend(warnings)
        sections.append(section)

    if not sections:
        inner = "_No skills installed. Run `/sf confirm` first._\n"
    else:
        inner = "# SkillForge skills\n\n" + "\n---\n\n".join(sections)

    block = f"{START_MARKER}\n{inner}\n{END_MARKER}"
    return block, all_warnings


# Matches the managed block with DOTALL so the inner markdown (which
# contains plenty of newlines) is captured entirely.
_MANAGED_RE = re.compile(
    re.escape(START_MARKER) + r".*?" + re.escape(END_MARKER),
    re.DOTALL,
)


def bundle_skills(
    skills: List[SkillInput],
    existing_agents_md: str = "",
) -> BundleResult:
    """Produce the full AGENTS.md content, preserving user content.

    Args:
        skills: Ordered list of skills to include. Order determines the
            section order in the output, so callers can sort by whatever
            makes sense (score, name, category).
        existing_agents_md: The current contents of AGENTS.md, or an
            empty string if the file doesn't exist.

    Returns:
        BundleResult with the merged content and any warnings collected
        from per-skill rendering.
    """
    managed_block, warnings = _render_managed_section(skills)

    if not existing_agents_md.strip():
        # Brand new file — just the managed block with a trailing newline
        content = managed_block + "\n"
        return BundleResult(content=content, warnings=warnings)

    if _MANAGED_RE.search(existing_agents_md):
        # Replace the existing managed section in place
        merged = _MANAGED_RE.sub(
            lambda _: managed_block,
            existing_agents_md,
            count=1,
        )
        if not merged.endswith("\n"):
            merged += "\n"
        return BundleResult(content=merged, warnings=warnings)

    # Existing file with no markers — append the managed block
    separator = "\n\n" if not existing_agents_md.endswith("\n\n") else ""
    appended = existing_agents_md.rstrip() + "\n\n" + managed_block + "\n"
    return BundleResult(content=appended, warnings=warnings)

File: src/export/codex.ts
/**
 * Codex CLI exporter — bundle skills into AGENTS.md managed section.
 *
 * Port of scripts/converters/codex.py.
 *
 * Strategy: managed section with HTML-comment markers.
 *   <!-- skillpro:start -->
 *   # SkillPro skills
 *   ... generated content ...
 *   <!-- skillpro:end -->
 *
 * User content lives outside the markers and survives re-runs.
 */

import { readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { homedir } from 'node:os';
import type { ScoredSkill } from '../types.js';

const START_MARKER = '<!-- skillpro:start -->';
const END_MARKER = '<!-- skillpro:end -->';

/**
 * Regex to match YAML frontmatter at the start of a file.
 */
const FRONTMATTER_RE = /^\s*---\s*\n([\s\S]*?)\n---\s*\n?/;

/**
 * Parse YAML frontmatter (reused from cursor converter logic).
 */
function parseFrontmatter(body: string): [Record<string, string>, string] {
  const match = FRONTMATTER_RE.exec(body);
  if (!match) return [{}, body];

  const fmText = match[1];
  const rest = body.slice(match[0].length);
  const fm: Record<string, string> = {};

  for (const line of fmText.split('\n')) {
    const kv = line.match(/^([A-Za-z_][A-Za-z0-9_-]*)\s*:\s*(.*)$/);
    if (!kv) continue;
    let val = kv[2].trim();
    if (val.length >= 2 && val[0] === val[val.length - 1] && (val[0] === '"' || val[0] === "'")) {
      val = val.slice(1, -1);
    }
    fm[kv[1]] = val;
  }

  return [fm, rest];
}

/**
 * Try to read the SKILL.md content for a given skill.
 */
function readSkillMd(skillId: string): string | null {
  const path = join(homedir(), '.claude', 'skills', skillId, 'SKILL.md');
  try {
    return readFileSync(path, 'utf-8');
  } catch {
    return null;
  }
}

/**
 * Render a single skill as a ## section for the managed block.
 */
function renderSkillSection(skill: ScoredSkill, mdContent: string): string {
  const [fm, body] = parseFrontmatter(mdContent);

  let description = (fm.description || '').trim();
  if (!description) {
    description = skill.id;
  }

  const bodyStripped = body.replace(/^\n+/, '').replace(/\n+$/, '');

  return (
    `## ${skill.id}\n` +
    `_Activate when: ${description}_\n\n` +
    `${bodyStripped}\n`
  );
}

/**
 * Build the full managed block bracketed by markers.
 */
function renderManagedSection(skills: ScoredSkill[]): string {
  const sections: string[] = [];

  for (const skill of skills) {
    const mdContent = readSkillMd(skill.id);
    if (!mdContent) {
      process.stderr.write(
        `  \u2718 ${skill.id}: SKILL.md not found, skipping Codex export\n`,
      );
      continue;
    }
    sections.push(renderSkillSection(skill, mdContent));
  }

  let inner: string;
  if (sections.length === 0) {
    inner = '_No skills installed. Run `/sf confirm` first._\n';
  } else {
    inner = '# SkillPro skills\n\n' + sections.join('\n---\n\n');
  }

  return `${START_MARKER}\n${inner}\n${END_MARKER}`;
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Regex to find the managed section.
 */
const MANAGED_RE = new RegExp(
  escapeRegex(START_MARKER) + '[\\s\\S]*?' + escapeRegex(END_MARKER),
);

/**
 * Export skills to AGENTS.md with managed section.
 *
 * Merge semantics:
 *   - If AGENTS.md exists with markers -> replace managed section
 *   - If AGENTS.md exists without markers -> append managed section
 *   - If no AGENTS.md -> create new file with managed section
 */
export function exportToCodex(
  skills: ScoredSkill[],
  projectDir: string,
): void {
  const agentsPath = join(projectDir, 'AGENTS.md');
  const managedBlock = renderManagedSection(skills);

  let existing = '';
  try {
    existing = readFileSync(agentsPath, 'utf-8');
  } catch {
    // File doesn't exist
  }

  let content: string;

  if (!existing.trim()) {
    // Brand new file
    content = managedBlock + '\n';
  } else if (MANAGED_RE.test(existing)) {
    // Replace existing managed section
    content = existing.replace(MANAGED_RE, managedBlock);
    if (!content.endsWith('\n')) {
      content += '\n';
    }
  } else {
    // Append managed section
    content = existing.trimEnd() + '\n\n' + managedBlock + '\n';
  }

  mkdirSync(dirname(agentsPath), { recursive: true });
  writeFileSync(agentsPath, content, 'utf-8');
  process.stderr.write(`  \u2713 AGENTS.md updated\n`);
}
